contracts/koh/contracts/KingAutomaton.sol

```diff
@@ -190,76 +211,135 @@ contract KingAutomaton {
+    BallotBox storage b = ballot_boxes[_id];
+    require(b.state == BallotBoxState.Active, "Ballot is not active!");
 
     // Calculate masks.
     uint index = _slot / VOTES_PER_WORD;
     uint offset = (_slot % VOTES_PER_WORD) * BITS_PER_VOTE;
     uint mask = BASE_MASK << offset;
 
     // Reduce the vote count.
-    uint vote = p.votes[index];
+    uint vote = b.votes[index];
     uint oldChoice = (vote & mask) >> offset;
     if (oldChoice > 0) {
-      p.voteCount[oldChoice]--;
+      b.voteCount[oldChoice]--;
     }
 
     // Modify vote selection.
     vote &= (mask ^ UINT256_MAX);   // get rid of current choice using a mask.
     vote |= _choice << offset;      // replace current choice using a mask.
-    p.votes[index] = vote;          // actually update the storage slot.
-    p.voteCount[_choice]++;         // update the total vote count based on the choice.
+    b.votes[index] = vote;          // actually update the storage slot.
+    b.voteCount[_choice]++;         // update the total vote count based on the choice.
 
     // Incentivize voters by giving them a refund.
-    p.payGas1[_slot] = 0;
-    p.payGas2[_slot] = 0;
-  }
+    b.payGas1[_slot] = 0;
+    b.payGas2[_slot] = 0;
 
-  function getVote(uint256 ballot_box_id, uint256 _slot) public view validBallotBoxID(ballot_box_id) returns (uint) {
-    BallotBox storage p = ballot_boxes[ballot_box_id];
+    // updateProposalState(_id);
+  }
 
+  function getVote(uint256 _id, uint256 _slot) public view validBallotBoxID(_id) returns (uint) {
     // Calculate masks.
     uint index = _slot / VOTES_PER_WORD;
     uint offset = (_slot % VOTES_PER_WORD) * BITS_PER_VOTE;
     uint mask = BASE_MASK << offset;
 
     // Get vote
-    return (p.votes[index] & mask) >> offset;
+    return (ballot_boxes[_id].votes[index] & mask) >> offset;
   }
 
-  function getVoteCount(uint256 ballot_box_id, uint256 _choice) public view validBallotBoxID(ballot_box_id) returns(uint256) {
-    return ballot_boxes[ballot_box_id].voteCount[_choice] & ALL_BUT_MSB;
+  function getVoteCount(uint256 _id, uint256 _choice) public view
+      validBallotBoxID(_id) returns(uint256) {
+    return ballot_boxes[_id].voteCount[_choice] & ALL_BUT_MSB;
+  }
+
+  function completeProposal(uint256 _id) private {
+    updateProposalState(_id);
+    BallotBox storage b = ballot_boxes[_id];
+    require(b.state == BallotBoxState.Active, "Ballot is not active!");
+    Proposal storage p = proposals[_id];
+    ProposalState p_state = p.state;
+    require(p_state == ProposalState.Started || p_state == ProposalState.Contested, "Invalid proposal state!");
+    p.state = ProposalState.Completed;
+    b.state = BallotBoxState.Inactive;
+  }
+
+  function updateProposalState(uint256 _id) public validBallotBoxID(_id) {
+    Proposal storage p = proposals[_id];
+    ProposalState p_state = p.state;
+    if (p_state == ProposalState.Uninitialized) {  // There is no proposal linked to this ballot
```

Comment:
Get rid of this check - if proposal is Uninitialized, then nothing is going to be done anyways.